ibis/common/caching.py

```diff
@@ -46,3 +56,67 @@ def __delitem__(self, key):
 
     def __repr__(self):
         return f"{self.__class__.__name__}({self._data})"
+
+
+class RefCountedCache:
+    """A cache with reference-counted keys.
+
+    We could implement `MutableMapping`, but the `__setitem__` implementation
+    doesn't make sense and the `len` and `__iter__` methods aren't used.
+
+    We can implement that interface if and when we need to.
+    """
+
+    def __init__(self, *, populate, lookup, finalize) -> None:
+        from ibis import util
+
+        self.cache = bidict()
+        self.refs = Counter()
+        self.populate = populate
```

Comment:
Perhaps it would be better to turn it into an abstract base class?

Reply:
Why? We wouldn't have any more than one child class.